### PR TITLE
perf(math): optimize radix2 domain

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR optimizes `CosetLDEBatch()` and `IFFTBatch()` in a high view.

### CODE REVIEW

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

File: tachyon/math/polynomials/univariate/radix2_evaluation_domain.h [hunk 1]
```diff
@@ -422,18 +422,20 @@ class Radix2EvaluationDomain : public UnivariateEvaluationDomain<F, MaxDegree>,
     }
     CHECK(base::bits::IsPowerOfTwo(mat.rows()));
     size_t cols = static_cast<size_t>(mat.cols());
-    size_t chunk_rows = size_t{1} << (this->log_size_of_group_ - mid_);
+    uint32_t log_n = this->log_size_of_group_;
+    uint32_t mid = log_n / 2;
+    size_t chunk_rows = size_t{1} << (this->log_size_of_group_ - mid);
```
Comment (GideokKim) [inline tachyon/math/polynomials/univariate/radix2_evaluation_domain.h:427]:
```suggestion
    size_t chunk_rows = size_t{1} << (log_n - mid);
```

File: tachyon/math/polynomials/univariate/radix2_evaluation_domain.h [hunk 2]
```diff
@@ -422,18 +422,20 @@ class Radix2EvaluationDomain : public UnivariateEvaluationDomain<F, MaxDegree>,
...
+    // max block size: 2^(|this->log_size_of_group_| - |mid|)
     OMP_PARALLEL_FOR(size_t block_start = 0; block_start < this->size_;
                      block_start += chunk_rows) {
       size_t thread = block_start / chunk_rows;
       size_t cur_chunk_rows = std::min(chunk_rows, this->size_ - block_start);
       Eigen::Block<RowMajorMatrix<F>> submat =
           mat.block(block_start, 0, cur_chunk_rows, cols);
-      for (uint32_t layer = mid_; layer < this->log_size_of_group_; ++layer) {
-        size_t first_block = thread << (layer - mid_);
+      for (uint32_t layer = mid; layer < this->log_size_of_group_; ++layer) {
```
Comment (GideokKim) [inline tachyon/math/polynomials/univariate/radix2_evaluation_domain.h:437]:
```suggestion
      for (uint32_t layer = mid; layer < log_n; ++layer) {
```

File: tachyon/math/polynomials/univariate/radix2_evaluation_domain.h [hunk 3]
```diff
@@ -99,12 +99,10 @@ class Radix2EvaluationDomain : public UnivariateEvaluationDomain<F, MaxDegree>,
       NOTREACHED();
     }
     CHECK_EQ(this->size_, static_cast<size_t>(mat.rows()));
-    uint32_t log_n = this->log_size_of_group_;
-    mid_ = log_n / 2;
 
     // The first half looks like a normal DIT.
     ReverseMatrixIndexBits(mat);
-    RunParallelRowChunks(mat, roots_vec_[log_n - 1], packed_roots_vec_[0]);
+    RunParallelRowChunks(mat, roots_vec_.back(), packed_roots_vec_[0]);
```
Comment (ashjeong) [inline tachyon/math/polynomials/univariate/radix2_evaluation_domain.h:105]: This appears belong to a separate commit (aka appears not relevant to moving `mid`.)

File: tachyon/math/polynomials/univariate/two_adic_subgroup.h [hunk 4]
```diff
@@ -29,12 +29,18 @@ class TwoAdicSubgroup {
     }
     FFTBatch(mat);
     Eigen::Index rows = mat.rows();
+    // TODO(chokobole): Change it to use |size_inv_| instead of directly
```
Comment (ashjeong) [inline tachyon/math/polynomials/univariate/two_adic_subgroup.h:33]:
```suggestion
    // TODO(chokobole): Use |size_inv_| instead of directly

File: tachyon/math/polynomials/univariate/two_adic_subgroup.h [hunk 5]
```diff
@@ -46,26 +47,34 @@ class TwoAdicSubgroup {
...
-        RowMajorMatrix<F>::Zero(rows << added_bits, cols));
+    RowMajorMatrix<F> new_mat(new_rows, cols);
+    OMP_PARALLEL_FOR(Eigen::Index row = 0; row < new_rows; ++row) {
+      if (row < rows) {
+        new_mat.row(row) = mat.row(row);
+      } else {
+        new_mat.row(row).setZero();
+      }
+    }
+    mat = std::move(new_mat);
```
Comment (ashjeong) [inline tachyon/math/polynomials/univariate/two_adic_subgroup.h:69]: shouldn't this be a separate commit like how https://github.com/kroma-network/tachyon/pull/521/commits/95a5d8886b963b5d0f14b7e383fa6a16556a97cd is?

Comment (chokobole) [reply tachyon/math/polynomials/univariate/two_adic_subgroup.h:69]: Looks like kind of mixed up. The motivation behind this change was that `conservativeResizeLike()` cannot be called with `Eigen::MatrixBase<Derived>`.

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Comment by batzor** [comment]
In ef58cee, I think you meant `refac(math): change CosetFFTBatch() to private` -> `refac(math): change CosetLDEBatch() to public`

File: tachyon/math/polynomials/univariate/radix2_evaluation_domain.h (hunk 3 above)
Comment (chokobole) [reply tachyon/math/polynomials/univariate/radix2_evaluation_domain.h:105]: Oh this is because computing `mid` depends on computing `log_n`, and `mid` was gone ad `log_n` was gone accordingly.

**Review by ashjeong — APPROVED** [review]
LGTM

**Comment by chokobole** [comment]
> In [ef58cee](https://github.com/kroma-network/tachyon/commit/ef58cee67e8adb01ec6a47ef850421f85876cfc5), I think you meant `refac(math): change CosetFFTBatch() to private` -> `refac(math): change CosetLDEBatch() to public`

No, the methods were originally all public, and I changed `CosetFFTBatch()` to private because it's only used internally. So the change was intentional and correct.

**Comment by batzor** [comment]
LGTM

**Review by batzor — APPROVED** [review]

**Review by GideokKim — APPROVED** [review]
LGTM
